opentimelineio/adapters/fcp_xml.py

```diff
@@ -562,95 +1375,129 @@ def _build_item_timings(
+    )
+    file_e.append(tc_element)
+
+    file_media_e = _get_or_create_subelement(file_e, "media")
+    if file_media_e.find("video") is None:
+        _append_new_sub_element(file_media_e, "video")
 
     return file_e
 
 
 @_backreference_build('file')
 def _build_file(media_reference, br_map):
-    file_e = cElementTree.Element('file')
+    file_e = _element_with_item_metadata("file", media_reference)
 
     available_range = media_reference.available_range
     url_path = _url_to_path(media_reference.target_url)
 
-    _insert_new_sub_element(file_e, 'name', text=os.path.basename(url_path))
-    file_e.append(_build_rate(available_range.start_time))
-    _insert_new_sub_element(
+    file_name = (
+        media_reference.name if media_reference.name
+        else os.path.basename(url_path)
+    )
+    _append_new_sub_element(file_e, 'name', text=file_name)
+    _append_new_sub_element(file_e, 'pathurl', text=media_reference.target_url)
+
+    # timing info
+    file_e.append(_build_rate(available_range.start_time.rate))
+    _append_new_sub_element(
         file_e, 'duration',
         text='{:.0f}'.format(available_range.duration.value)
     )
-    _insert_new_sub_element(file_e, 'pathurl', text=media_reference.target_url)
 
     # timecode
-    timecode = available_range.start_time
-    timecode_e = _insert_new_sub_element(file_e, 'timecode')
-    timecode_e.append(_build_rate(timecode))
-    # TODO: This assumes the rate on the start_time is the framerate
-    _insert_new_sub_element(
-        timecode_e,
-        'string',
-        text=otio.opentime.to_timecode(timecode, rate=timecode.rate)
-    )
-    _insert_new_sub_element(
-        timecode_e,
-        'frame',
-        text='{:.0f}'.format(timecode.value)
+    ref_tc_metadata = media_reference.metadata.get(META_NAMESPACE, {}).get(
+        "timecode"
     )
-    display_format = (
-        'DF' if (
-            math.ceil(timecode.rate) == 30 and math.ceil(timecode.rate) != timecode.rate
-        ) else 'NDF'
+    tc_element = _build_timecode_from_metadata(
+        available_range.start_time, ref_tc_metadata
     )
-    _insert_new_sub_element(timecode_e, 'displayformat', text=display_format)
+    file_e.append(tc_element)
 
     # we need to flag the file reference with the content types, otherwise it
     # will not get recognized
-    file_media_e = _insert_new_sub_element(file_e, 'media')
-    content_types = []
-    if not os.path.splitext(url_path)[1].lower() in ('.wav', '.aac', '.mp3'):
-        content_types.append('video')
-    content_types.append('audio')
+    # TODO: We should use a better method for this. Perhaps pre-walk the
+    #       timeline and find all the track kinds this media is present in?
+    if not file_e.find("media"):
+        file_media_e = _get_or_create_subelement(file_e, "media")
 
-    for kind in content_types:
-        _insert_new_sub_element(file_media_e, kind)
+        audio_exts = {'.wav', '.aac', '.mp3', '.aif', '.aiff', '.m4a'}
```

Comment:
Just so I understand, FCP doesn't explicitly mark a track as audio or video?  Its implicit based on what the thing is pointed at?

Reply:
In the Final Cut XML, the structure of the tags for tracks basically look like 
```
<project> <children> <sequence> 
    <media> <video>
            <track/> <track/> ...
    </video>
    <audio>
            <track/> <track/> ...
    </audio> </media>
</sequence> </children> </project>
```

where "sequence" == "a timeline", and a particular track element is in either a "video" or "audio" element, and only has one kind of media one it.  (But one _file_ can wind up referenced in both a video and audio track, linked so you can drag them as a unified thing in the UI of something like Premiere.)  If you use OTIO to populate only the video track with a quicktime file, it shows up as silent.  Likewise, if you put a .wav file in a video track, it won't work right, so needs to be filtered out from video tracks.

All that said, this spot in fcp_xml.py is for making a file, which can potentially exist outside of a track, and sort of has its own semantics, because xmeml wants to rot your brain.

When a video <track> has a <clipitem>, the clipitem has a <file> which has its own <media> element that has to have both <video> and <audio> elements describing the format if the actual file does -- even if the track that contains the element is only a video track.  I have seen Premiere fail to link a Quicktime file if the XML omits the the <audio> tag, even if it is only in a <video> track, because in sees the file has two audio channels, but it thinks the file referenced in the timeline is supposed to have none, and thinks there is a mismatch.

I _think_ that answers your question, but I fear I may have answered your question by making it even less clear.  :)

Reply:
Also, this whole area of logic feels a bit CBB to me. However, the solve in unclear to me at the moment.